lxd/main_callhook.go

```diff
@@ -24,14 +32,166 @@ func (c *cmdCallhook) Command() *cobra.Command {
   Call container lifecycle hook in LXD
 
   This internal command notifies LXD about a container lifecycle event
-  (start, stopns, stop, restart) and blocks until LXD has processed it.
+  (start, startmountns, stopns, stop, restart) and blocks until LXD has processed it.
 `
 	cmd.RunE = c.Run
 	cmd.Hidden = true
 
+	// devicesRootFolder is used to specify where to look for CDI config device files.
+	cmd.Flags().StringVar(&c.devicesRootFolder, "devicesRootFolder", "", "Root folder for CDI devices")
+
 	return cmd
 }
 
+// resolveTargetRelativeToLink converts a target relative to a link path into an absolute path.
+func resolveTargetRelativeToLink(link string, target string) (string, error) {
+	if !filepath.IsAbs(link) {
+		return "", fmt.Errorf("The link must be an absolute path: %s (target: %s)", link, target)
+	}
+
+	if filepath.IsAbs(target) {
+		return target, nil
+	}
+
+	linkDir := filepath.Dir(link)
+	absTarget := filepath.Join(linkDir, target)
+	cleanPath := filepath.Clean(absTarget)
+	absPath, err := filepath.Abs(cleanPath)
+	if err != nil {
+		return "", err
+	}
+
+	return absPath, nil
+}
+
+// linuxArchTriplet matches the architecture triplet of a Linux system.
+// e.g. x86_64-linux-gnu, aarch64-linux-gnueabihf, etc.
+var linuxArchTriplet = regexp.MustCompile(`([\w-]+)-linux-([\w-]+)`)
+
+// getArchTripletFromPaths returns the architecture triplet from a list of paths.
+func getArchTripletFromPaths(paths []string) (string, error) {
+	for _, path := range paths {
+		matches := linuxArchTriplet.FindStringSubmatch(path)
+		if len(matches) > 2 {
+			return matches[1] + "-linux-" + matches[2], nil
+		}
+	}
+
+	return "", fmt.Errorf("Could not determine architecture triplet from paths")
+}
+
+// onStartMountNS is a special hook that is called before the container has started but after the container
+// namespace has been setup. This hook executes a series of operation (described in a 'hooks file'),
+// when a CDI device need to be passed to an instance.
+func onStartMountNS(devicesRootFolder string, hooksFilePath string) error {
```

Comment:
Lets make this function name mention CDI please

Reply:
fixed